lib/sqlalchemy/orm/properties.py

```diff
@@ -335,3 +358,14 @@ def reverse_operate(self, op, other, **kwargs):
 
     def __str__(self):
         return str(self.parent.class_.__name__) + "." + self.key
+
+
+class _ColumnPropertyProxy(ColumnProperty):
```

Comment:
why do we need a new class?

Reply:
This new class is used by the `public_factory` function of `column_property` to set the default value of the parameter `raise_on_set` to `"warn"` to raise a warning regarding the change in default behaviour of a computed attribute. This class would then be used to change the default value to True to enable the new behaviour by default.
ColumnProperty should always keep its default to False.

Ideally the ability to set a default attribute should probably be on `public_factory`, but it does not support it currently.

As I wrote here https://github.com/sqlalchemy/sqlalchemy/pull/4972#discussion_r344456212 I don't like this implementation, but I don't have another alternative without changing public_factory

Reply:
public_factory can refer to any classmethod or callable so if we wanted to have two different defaults for "raise_on_set", that would be achieved by putting a `@classmethod` factory on `ColumnProperty`.

However, whether or not a `column_property()` is able to receive a set event is not actually in the user's control.   If the property refers to one of the columns that is mapped to the table, as you've observed is how it is used internally, then the attribute should allow set events to occur.     But the way `column_property()` is used internally, that is, as a wrapper around a `Column` that is mapped, is also [public api](https://docs.sqlalchemy.org/en/13/orm/mapping_columns.html#using-column-property-for-column-level-options).

so while the flag here should be present at the attribute level, from the `ColumnProperty` perspective, it likely needs to inspect the expression given in terms of the parent `Mapper` and determine this value dynamically.   This implies some level of reverse-implementing some of what happens in https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/orm/persistence.py#L569, which is determining if the expression mapped is directly the same as `mapper._propkey_to_col[table]`  where 'table' is also present in `mapper._pks_by_table`.  this could be encapsulated in a method on `Mapper` such as `_expression_is_persistent(expr)` which tests that the expression is one of the mapped columns on one of the persistent mapped tables.

Reply:
> public_factory can refer to any classmethod or callable so if we wanted to have two different defaults for "raise_on_set", that would be achieved by putting a @classmethod factory on ColumnProperty.

that makes more sense

> But the way column_property() is used internally, that is, as a wrapper around a Column that is mapped, is also public api.

I had not realized this. I thought its use was just to map an sql expression, similar to a computed column. My implementation does not work for other use cases.

I'm not sure I can complete this one in a satisfactory way, since there are a lot more cases that I'm not familiar with. 😞

Reply:
I can close this and open an issue instead, since this implementation looks like a dead end

Reply:
it's not too difficult but we should likely have an issue to track the "warn for writing to non-persisted columns" behavior.    it's known at mapping time which attributes will be persistable and which will not.

Reply:
>  it's known at mapping time which attributes will be persistable and which will not.

is there a property with this information that I could use to set the behavior in column attribute?

Reply:
no, but im proposing we add one like this to mapper.py.  

looking at where this happens, I think we should just have the mapper set a flag directly on the property, as we have determined if an incoming column_property() is readonly right here: https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/orm/mapper.py#L1718 and also here: https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/orm/mapper.py#L1403.

So there's a little bit of chicken/egg stuff happening with this as these column properties can be added to existing mappers or new mappers and things might happen in different orders, but overall, once the ColumnProperty is all set up, if it were to ask the mapper, "self in self.mapper._readonly_props", that's the correct answer.

Reply:
will try to work something with it.
are computed column already treated as readonly?

Reply:
that seems to be the case, yes.   this collection is used to determine properties that we're going to refresh automatically after a flush but it's also the properties for which we aren't going to persist anything.

Reply:
A computed column seem to not go in `_readonly_props` 

```py
Base = declarative_base()
class Foo(Base):
    side= sa.Column(sa.Float, primary_key=True)
    area = sa.Column(sa.Float, sa.Computed(side*side))
    test = sa.orm.column_property(area + 33)
    __tablename__='foo'
Foo.area.prop in Foo.area.parent._readonly_props # False
Foo.test.prop in Foo.test.parent._readonly_props # True
```

Reply:
ah well I think it should.  I can see why it doesn't, because it looks like a regular column.  The mapper should add a test to see if the column acts as "computed" when it determines this collection.   we might want to put a flag on Column that makes this simple, like "_is_readonly" or something.  note the underscore is so we don't have to worry about end users interpreting this as something else.

Reply:
I guess it depends on the meaning of `_readonly_props` like I wrote here https://github.com/sqlalchemy/sqlalchemy/pull/4972#issuecomment-565815472 

I agree that a property in the column would be useful, so the it can be generalized.